File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/AgentFOV.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * AgentFOV - Update to agent's field of view, angle is vertical, single F32 float in radians
 *
 * <p>Template: {@code AgentFOV Low 82 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
public class AgentFOV extends SLMessage {
    public AgentData AgentData_Field;
    public FOVBlock FOVBlock_Field;

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
        public int CircuitCode; // U32
        public UUID SessionID; // LLUUID
    }

    /** Block FOVBlock, Single. */
    public static class FOVBlock {
        public int GenCounter; // U32
        public float VerticalAngle; // F32
    }

    public AgentFOV() {
        this.zeroCoded = false;
        this.AgentData_Field = new AgentData();
        this.FOVBlock_Field = new FOVBlock();
    }

    @Override
    public int CalcPayloadSize() {
        return 48;
    }

    @Override
    public void Handle(SLMessageHandler messageHandler) {
        messageHandler.HandleAgentFOV(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 82 (AgentFOV).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x52);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        packInt(byteBuffer, this.AgentData_Field.CircuitCode);
        packInt(byteBuffer, this.FOVBlock_Field.GenCounter);
        packFloat(byteBuffer, this.FOVBlock_Field.VerticalAngle);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        this.AgentData_Field.CircuitCode = unpackInt(byteBuffer);
        this.FOVBlock_Field.GenCounter = unpackInt(byteBuffer);
        this.FOVBlock_Field.VerticalAngle = unpackFloat(byteBuffer);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/AgentFOV.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * AgentFOV - Update to agent's field of view, angle is vertical, single F32 float in radians
 *
 * <p>Template: {@code AgentFOV Low 82 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
open class AgentFOV : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var FOVBlock_Field: FOVBlock = FOVBlock()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var CircuitCode: Int = 0
        @JvmField var SessionID: UUID? = null
    }

    /** Block FOVBlock, Single. */
    open class FOVBlock {
        @JvmField var GenCounter: Int = 0
        @JvmField var VerticalAngle: Float = 0f
    }

    init {
        zeroCoded = false
    }

    override fun CalcPayloadSize(): Int {
        return 48
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleAgentFOV(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 82 (AgentFOV).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x52).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        packInt(byteBuffer, AgentData_Field.CircuitCode)
        packInt(byteBuffer, FOVBlock_Field.GenCounter)
        packFloat(byteBuffer, FOVBlock_Field.VerticalAngle)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        AgentData_Field.CircuitCode = unpackInt(byteBuffer)
        FOVBlock_Field.GenCounter = unpackInt(byteBuffer)
        FOVBlock_Field.VerticalAngle = unpackFloat(byteBuffer)
    }
}
